core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -505,13 +497,13 @@ class KafkaApis(val requestChannel: RequestChannel,
         BrokerTopicStats.getBrokerAllTopicsStats().bytesOutRate.mark(data.records.sizeInBytes)
       }
 
-      val response = new FetchResponse(versionId, fetchedPartitionData, 0)
+      val response = new FetchResponse(fetchedPartitionData, 0)
+      val responseStruct = response.toStruct(versionId)
 
-      def fetchResponseCallback(delayTimeMs: Int) {
-        trace(s"Sending fetch response to client $clientId of " +
-          s"${convertedPartitionData.map { case (_, v) => v.records.sizeInBytes }.sum} bytes")
-        val fetchResponse = if (delayTimeMs > 0) new FetchResponse(versionId, fetchedPartitionData, delayTimeMs) else response
-        requestChannel.sendResponse(new RequestChannel.Response(request, fetchResponse))
+      def fetchResponseCallback(throttleTimeMs: Int) {
+        trace(s"Sending fetch response to client $clientId of ${responseStruct.sizeOf} bytes.")
+        val responseSend = response.toSend(responseStruct, throttleTimeMs, request.connectionId, request.header)
+        requestChannel.sendResponse(new RequestChannel.Response(request, responseSend))
```

Comment:
@apurvam Good catch, I counted two Structs from `FetchRequest` creation (if `throttleTimeMs > 0`) and one from `toSend`, but the latter didn't happen before my change. So, we just save a `Struct` creation if `throttleTimeMs > 0`.